build_tools/buildkite/scripts/update_pipeline_configuration.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python3
+# Copyright 2022 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+"""Updates the gievn pipelines with the Buildkite API.
+
+This overwrites the YAML steps of all the Buildkite pipeline. Pipelines should
+be updated this way, not in the UI. By default before updating, this checks if a
+later job with the same step key has already run successfully, in which case it
+skips the update. This is to avoid situations where an earlier build is for some
+reason running a step after a later build due to race conditions or a retry.
+Buildkite concurrency groups should be used to prevent builds from running this
+check simultaneously.
+"""
+
+import argparse
+import os
+from pybuildkite import buildkite
+import requests
+import sys
+
+PIPELINE_ROOT_PATH = "build_tools/buildkite/pipelines"
+
+
+def parse_args():
+  parser = argparse.ArgumentParser(
+      description="Updates the YAML steps of a Buildkite pipeline.")
+  parser.add_argument(
+      "pipeline",
+      help=(f"The slug of the pipeline to update. The steps are pulled from the"
+            f" file at {os.path.join(PIPELINE_ROOT_PATH, '<pipeline>.yml')}"))
+  return parser.parse_args()
+
+
+def should_update(bk, organization, pipeline_to_update):
+  # Buildkite sets these environment variables. See
+  # https://buildkite.com/docs/pipelines/environment-variables. If running
+  # locally you can set locally or use the simulate_buildkite.sh script.
+  running_pipeline = os.environ["BUILDKITE_PIPELINE_SLUG"]
+  build_number = int(os.environ["BUILDKITE_BUILD_NUMBER"])
+  step_key = os.environ["BUILDKITE_STEP_KEY"]
+
+  next_build_number = build_number + 1
```

Comment:
I don't think that's possible based on the docs: https://buildkite.com/docs/pipelines/environment-variables#bk-env-vars-buildkite-build-number. That doesn't give extremely strong guarantees, but given that keeping a counter going across builds seems not that hard, I bet they've implemented this such that it's always true. I looked into pipeline-level metadata and unfortunately there doesn't seem to be any. I could fetch all builds for a pipeline and then filter, but that's a lot more fiddly (have to deal with pagination and such), so it didn't seem worth it.

Reply:
Does it prevent the following case? Assume we have three builds A, B, C with the build number 10, 11, 12
The build B and C finished before A runs this check, and B failed but C succeed. When A does this check, it will check the build B and know it failed, so A will update the pipeline config and overwrite the later build C.

This probably happens extremely rarely tho...

Reply:
Aw damn, you're right. Ugh I really do want pipeline metadata here somewhere. hmm... I may have to think on this a bit more or contact support to see if there's a solution. You're right that this would be exceedingly rare (somehow builds 11 and 12 would've had to start before build 10), but I don't like messing about with race conditions. I could embed the metadata about which build did the update in a comment in the steps themselves. Totally a hack, but not the worst one in the world

Reply:
> I could embed the metadata about which build did the update in a comment in the steps themselves

I had something like that in mind when viewing 
![image](https://user-images.githubusercontent.com/4010439/159973316-3e8c6323-1213-47e4-939b-777d51c346b2.png)

and writing this comment: https://github.com/google/iree/pull/8606#discussion_r832393902

If the steps had a "automatically imported from [script url] at [build number]" comment at the top it would explain things pretty nicely.

Reply:
Thanks for catching this @pzread! Concurrency is hard and I really appreciate having thorough code reviews :slightly_smiling_face:

Reply:
Fixed